packages/idempotency/src/middleware/makeHandlerIdempotent.ts

```diff
@@ -0,0 +1,144 @@
+import { IdempotencyHandler } from '../IdempotencyHandler';
+import { IdempotencyConfig } from '../IdempotencyConfig';
+import { cleanupMiddlewares } from '@aws-lambda-powertools/commons/lib/middleware';
+import {
+  IdempotencyItemAlreadyExistsError,
+  IdempotencyPersistenceLayerError,
+} from '../Exceptions';
+import { IdempotencyRecord } from '../persistence';
+import type {
+  MiddlewareLikeObj,
+  MiddyLikeRequest,
+} from '@aws-lambda-powertools/commons';
+import type { IdempotencyLambdaHandlerOptions } from '../types';
+
+/**
+ * A middy middleware to make your Lambda Handler idempotent.
+ *
+ * @example
+ * ```typescript
+ * import {
+ *   makeHandlerIdempotent,
+ *   DynamoDBPersistenceLayer,
+ * } from '@aws-lambda-powertools/idempotency';
+ * import middy from '@middy/core';
+ *
+ * const dynamoDBPersistenceLayer = new DynamoDBPersistenceLayer({
+ *   tableName: 'idempotencyTable',
+ * })
+ *
+ * const lambdaHandler = async (_event: unknown, _context: unknown) => {
+ *   //...
+ * };
+ *
+ * export const handler = middy(lambdaHandler)
+ *  .use(makeHandlerIdempotent({ persistenceStore: dynamoDBPersistenceLayer }));
+ * ```
+ *
+ * @param options - Options for the idempotency middleware
+ */
+const makeHandlerIdempotent = (
+  options: IdempotencyLambdaHandlerOptions
+): MiddlewareLikeObj => {
+  const idempotencyConfig = options.config
+    ? options.config
+    : new IdempotencyConfig({});
+  const persistenceStore = options.persistenceStore;
+  persistenceStore.configure({
+    config: idempotencyConfig,
+  });
+
+  /**
+   * Function called before the handler is executed.
+   *
+   * Before the handler is executed, we need to check if there is already an
+   * execution in progress for the given idempotency key. If there is, we
+   * need to determine its status and return the appropriate response or
+   * throw an error.
+   *
+   * If there is no execution in progress, we need to save a record to the
+   * idempotency store to indicate that an execution is in progress.
+   *
+   * @param request - The Middy request object
+   */
+  const before = async (request: MiddyLikeRequest): Promise<unknown | void> => {
```

Comment:
Currently the middleware doesn't implement the retry mechanism that is [present in `IdempotencyHandler.handle`](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/244cfd2b42d24318ab1d149ab694c68109aef342/packages/idempotency/src/IdempotencyHandler.ts).

I am not 100% sure whether this can be done in the context of a middleware and before being sure I need to better understand when is that we can/need retry. Could we have a quick chat so maybe you can help me?

Reply:
The retry mechanism is already handled by `IdempotencyHandler`, so there's no additional implementation to do here.